### DISCUSSION TITLE
host crashing on IPlugView::attached() call to IPlug2 vst3s

### DISCUSSION THREAD

**Discussion by cuikp** [iPlug2/iPlug2#1105] [Q&A]
I have a custom (cross-platform) host, performing very simple operations for initializing and displaying vst3s.  It works well for most vst3s I've tested -- except for Vst3s based on IPlug2.  On Windows, these plugins will crash every time I attempted to call IPlugView::attached(windowHandle, platform), producing a fatal memory access error which can't be caught by try-catch block by my C++ host.  (On MacOS, the .dylib versions of the same vst3s display properly without issue.)

After downloading the IPlugEffect.vst3 with the associated .pdb file I've been able to trace the call stack and method that produces the error:
Exception thrown: read access violation.
iplug::igraphics::StaticStorage<iplug::igraphics::APIBitmap>::Accessor::Find(...) returned nullptr.

with the very last call shown on the call stack being:
 
IPlugEffect.vst3!iplug::igraphics::IGraphicsSkia::PrepareAndMeasureText(const iplug::igraphics::IText & text, const char * str, iplug::igraphics::IRECT & r, double & x, double & y, SkFont & font) Line 512	C++

but since that is internal to IPlug2, I'm not sure how my host can circumvent it.

My IEditController, IPlugView, IPlugFrame etc. pointers, windowHandle and platform ("HWND") are most definitely valid (work for most vsts) and even with IPlug2, the expected properties are being returned, such as setFrame(), IsPlatformSupported() and getSize(ViewRect) (correct width/height), just prior to the crash.  

Yet I'm obviously failing to provide IPlug2 with some information it expects, which other vst3s are not expecting, but what might that be?

**Comment by AlexHarker** [comment]
This shouldn't happen if you are using IPlugEffect.vst3 from the website - have you tested it in any other windows host?

This also doesn't look like you not doing anything correctly, although it is a bit hard to tell.

However, I don't see how you'd get an exception in that Find() method from PrepareAndMeasureText() because it doesn't call that. If we take it that the exception is correct then it looks like somewhere the graphics backend tries to create a bitmap that fails.

The first thing to try is getting hold of a build (or making one yourself) of iPlugEffect.vst3 that uses NanoVG (which should be the default. Or if you can share your host I can try to take look?

**Reply by cuikp** [reply]
Yes I also think that's a strange place for an error to occur (PrepareAndMeasureText) -- but that could just be the last line that was able to be logged on the call stack.  

My host is a rather large project, so I've prepared a barebones code extract of just the portion that produces the error (with IPlug2 vst3s).  Here's a link to a zip with both the .exe and .pdb.  If you need I can provide the code for it as well.

https://lavenderlandweb.com/freeFiles/StraightCHost.zip

This is setup to be very simple - choose vst3 directory, select a vst3 to show vstinfo and initialize the plugin.
(even IPlug2 vst3s initialize without problem).  
Once initialized, press the "Show editor" button (this crashes IPlug2-based vst3s).

Incidentally, the crash doesn't occur *exclusively* with IPlug2s (for example I get a similar although not identical error for Steinberg's Backbone), though most other vst3s such as AddictiveDrums, Auburn Sounds Graillon, numerous others, load their editors fine.

Thanks for any help or suggestions you might have!

**Comment by AlexHarker** [comment]
OK - I can confirm a crash with another IPlug2 plugin, but I need to do some updates on my windows system to be able to investigate further - I'll see if I can get more clarity.
